### ISSUE TITLE
Interface support

### ISSUE THREAD

**Issue by inf3rno** [inf3rno/o3#104]
I decided to add some interface support to the code.  I came up with a few possible solutions:
http://stackoverflow.com/questions/41788679/domain-driven-design-for-node-js/42418735#42418735

There is another implementation with getOwnPropertyDescriptors.
https://www.webreflection.co.uk/blog/2016/03/23/javascript-interfaces

The ideal would be to use a polyfill with both toString and getOwnPropertyDescriptors. So it would be more or less backward compatible.

I am not sure whether I want to put this feature into the class lib. I'll think about this. I did not want to be too oo by implementing this lib. So no interfaces, method override, type check, etc...

**Comment by inf3rno** [comment]
To add interface support to classes we need to wrap the constructor. So by instantiating the class we can throw an abstract class instantiation exception. Wrapping the constructor is not a good idea, because I won't be able to extend already existing classes with the class related methods. So I am pretty sure that this approach won't work. I don't want to wrap methods either and I don't want to have auto type check on the arguments.

I need a different approach.

**Comment by inf3rno** [comment]
The upper means that I cannot verify the interface by instantiation. All I can do is verifying the interface by adding it to the class.

```js
var i = Interface.extend({
    a: {x: String}
});
var j = Interface.extend({
    b: {}
});
var c = Class.extend({
    prototype: {
        a: function (x){}
    }
});

c.implements(i);
// ok

c.implements(j);
// error: b() is not implemented
```

There is cache, so if `c` implements `i`, then it will be cached until `c` is modified by a class method. Modification of `c` invalidates the cache. We can check next time when asked, or we can set this rule permanent. I prefer the latter one. Which means after `c.implements(i)` is called, `c` and its descendants MUST implement `i` at any time.

```js
var i = Interface.extend({
    a: {x: String}
});
var c = Class.extend({
    prototype: {
        a: function (x){}
    }
});

c.implements(i);
// ok

c.prototype.merge({
    a: function (x, y){}
});
// error: a(x,y) does not fulfill the a(x) interface constraint
```

Interfaces must be immutable, so we can define them only once and they cannot be modified after that. Interfaces can have a single ancestor interface, but the descendant interfaces cannot override or make constraints on the methods defined by the ancestor interface.

```js
var i = Interface.extend({
    a: {x: String}
});

var j = i.extend({
    b: {y: Number}
});
// ok

var k = i.extend({
    a: {y: Number}
});
// error: a(y) does not fulfill the a(x) interface constraint
// error: cannot redeclare the a(x) interface constraint (maybe this latter one is better)
```

Type check?

I am certain I won't wrap any method, but it doesn't hurt if we provide type checking solutions based on the interface. That means:

```js
var i = Interface.extend({
    a: {x: String}
});
var c = Class.extend({
    prototype: {
        a: function (x){
            this.validate({a: {x:x}});
            // or
            this.validate({a: [x]});
            // or
            this.validate(this.a, arguments);
            // probably the latter one requires the least effort
        }
    }
});

c.implements({
    i:i
});
// alternative approach to name interfaces locally
// not necessary, it might be better to name them the same way we did by the classes
```

**Comment by inf3rno** [comment]
I think that's all related to interfaces and validation. It can be in a separate project as well, maybe i3. I am not sure yet. I have to override merge and absorb for certain, because I need the verification code to put there.

**Comment by inf3rno** [comment]
Another possible approach to declare interfaces without type checking:

```js
var i = new Interface({
    a: function (x){}
});
```

This can be okay as long as we don't need type checking. Maybe we can put the type checking code into this function.

```js
var i = new Interface({
    a: function (x){
        x.implements(j);
    }
});
```

**Comment by inf3rno** [comment]
What if we don't try to implement type check and validation based on this? I'd like to keep it simple, so type check should be added to the actual method as usual:

```js
Q = Class.extend({
    prototype: {
        m: function (x){
            if (!(x instanceof Y))
                throw new InvalidArgumentException("...");
            ...
        }
    }
});
```

So using interface does not do type check, it just check whether the actual methods are implemented.

```js
iQ = new Interface({
    prototype: {m: function (x){}}
});
Q = Class.extend({
    prototype: {
        m: function (x){
            if (!(x instanceof Y))
                throw new InvalidArgumentException("...");
            ...
        }
    }
}).implement(iQ);
```

or probably 


```js
iQ = new Interface({
    prototype: {m: function (x){}}
});
Q = Class.extend({
    implement: iQ,
    prototype: {
        m: function (x){
            if (!(x instanceof Y))
                throw new InvalidArgumentException("...");
            ...
        }
    }
})
```

or maybe

```js
iQ = new Interface({
    prototype: {m: function (x){}}
});
Q = Class.extend(iQ, {
    prototype: {
        m: function (x){
            if (!(x instanceof Y))
                throw new InvalidArgumentException("...");
            ...
        }
    }
})
```

I like this latter one. The only problem that it is not compatible with instanceof, but who cares. :-)

Afaik. interfaces define only methods, but not static methods. So I should not use prototypes in them.

```js
iQ = new Interface({
    m: function (x){}
});
```

Ofc. we don't have to stick to the old definition, but I think it is fine this way.

This is all good, but I designed this lib to never override the constructor, so it is compatible with every other existing lib. I don't want to give this feature up just to check whether the instantiated class is fully implementing an interface.

I think that was the cause I decided to support only on the fly checks of instances. e.g.

```js
iQ = new Interface({
    m: function (x){}
});
Q = Class.extend({
    prototype: {
        m: function (x){
            if (!(x instanceof Y))
                throw new InvalidArgumentException("...");
            ...
        }
    }
});
P = Class.extend({
    prototype: {
        a: function (q){
            if (!q.implements(iQ))
                throw new InvalidArgumentException("...");
            ...
        }
    }
});
```
I think this should be the final format. We can add a cache for classes and instances, so using toString by methods and checking the argument list would be necessary only for a single time.

**Comment by inf3rno** [comment]
I think we can add support for this on the fly check. I added it to the task list, but it is not a priority task and it should be added only with a proper caching mechanism, otherwise it would be too slow.

**Comment by inf3rno** [comment]
```js
iQ = new Interface({
    m: function (x){}
});
Q = Class.extend({
    prototype: {
        m: function (x){
            if (!(x instanceof Y))
                throw new InvalidArgumentException("...");
            ...
        }
    }
});
P = Class.extend({
    prototype: {
        a: function (q){
            if (!q.implements(iQ))
                throw new InvalidArgumentException("...");
            ...
        }
    }
});
```

Let's check what happens here:

 - `q.implements(iQ)` is called - probably not the best naming, but ok for now
 - iterates through the methods defined in iQ
 - check where these methods was defined by q
 - if it was defined in an ancestor, then check the implementation of the ancestor for every method
 - if one of the methods are not implemented or is an abstractMethod, then it steps to the previous ancestor (which is closer to the actual instance)
 - the cache meanwhile logs whether a method is implemented in a class
 - if it doesn't find a class, which implements every methods, then it checks the intance
 - if the instance implements the missing methods, then it is fine, otherwise, the result will be false

The cache is an interesting topic here. We can do `method <-> interface.method` cache and we can do `class <-> interface` cache. I think we should do both, so if we have an abstract class and the instance implements the last method, then the other methods of the class won't be checked again and again. Method level cache guarantees that. Class level cache makes checking a class somewhat faster, but it is probably not as important as method level cache, because there is no string comparison involved there.

I saw somewhere that the arguments of a method can be listed, but I think that was not ES6, but some kind of transpiled stuff, probably typescript. I'll investigate it.

**Comment by inf3rno** [comment]
I saw somewhere that the arguments of a method can be listed, but I think that was not ES6, but some kind of transpiled stuff, probably typescript. I'll investigate it.

If I don't find it, then we can do something like this (not sure about the regex flags and the match output):

```js
Function.prototype.listArguments = function listArguments(){
    if (this.args)
        return this.args;
    var str = this.toString();
    var pattern = /^function\s+\w+\s*\(([^)]*)\)/;
    var args = pattern.match(str)[1].replace(" ","").split(",");
    if (args[0] === "")
        args.length = 0;
    this.args = args;
    return this.args;
};
```
This is fine, if I lazy load this when it is needed.

**Comment by inf3rno** [comment]
Should I allow extension by the arg list or should it be a exact match?
I think it is simpler if it is an exact match and probably more reliable as well.

```js
iQ = new Interface({m: function (x,y){}});
({m: function (x){}}).implements(iQ); //false
({m: function (x,y){}}).implements(iQ); //true
({m: function (x,y,z){}}).implements(iQ); //false
({m: function (x,z,y){}}).implements(iQ); //false
```

**Comment by inf3rno** [comment]
I think the `o.implements(i)` is a bad solution, because I have to override the `Object.prototype`. What are the other options?

`iQ.implementedby(o)` - fine, but cannot check multiple interfaces at once (probably no need for that though most of the times
`Interface.implementedby.call(iQ,o)` - fine
`Interface(iQ,iP).implementedby(o1,o2)`

I think all of these are fine and we can do something like 
`Interface.implementedby(o,iQ)` too, because we can check whether something has an interface type

I think it is important that `iQ instanceof Interface` should be true and we should be able to inherit interfaces with extend in a similar way as classes.

**Comment by inf3rno** [comment]
ES6 supports some kind of enumeration: `function dumpArguments(...args) {` I think it would be better to normalize (remove whitespace) and check for string match instead of trying to parser everything. It is compatible with the new features. I need a HashMap(f, argsString) for that. That is cool if we have HashMap support, otherwise it can't be done without a memory leak.

**Comment by inf3rno** [comment]
#105 (WeakMap polyfill needed for cache)

**Comment by inf3rno** [comment]
It is possible to support ES6 features in the argument list without parsing it. At least I think so. I need to look for `(` and `)` chars so I will know when the argument list is ended. The only problem when these chars appear in strings, so I guess I have to look for strings as well. Still better than including a complete js parser. I need to check exactly what is supported in ES6 argument lists. Maybe it is a lot easier than I think.
